VueJS/index.js

```diff
@@ -1,13 +1,84 @@
-import List from './components/List.js'
-import ListItem from './components/ListItem.js'
+Vue.component('todo-list', {
+	template: `<div>
+		<ul>
+			<todo-item
+				v-for="(todo, index) in todos"
+				v-bind:todo="todo"
+				v-on:save="$emit('save', {index: index, message: $event})"
+				v-on:remove="$emit('remove', index)"
+			/>\
+		</ul>
+		<button v-on:click="$emit('add')">Add new todo</button>
+		</div>`,
+	props: ['todos']
+})
+
+Vue.component('todo-item', {
+  template: `<li>
+	<template v-if="editing">
+		{{ todo.id }} {{ todo.message }}
+		<button v-on:click="editing=false">Edit</button>
+	</template>
+	<template v-else>
+		<input placeholder="Type todo here..." v-model="input"/>
+		<button v-on:click="save">Save</button>\
+		<button v-on:click="editing=true">Cancel</button>
+	</template>
+	<button v-on:click="remove">Remove</button>
+    </li>`,
+  props: ['todo'],
+  data: function() {
+	return {
+		editing: !(this.todo.message === ""),
```

Comment:
I think I have seen this before. I don't like this solution of creating an empty todo in order to have the edit form open. Why not set this to `true` by default and maybe override it with a prop? Or simply create a new todo in one go instead. That way, you don't have this in-between state.